src/zap/zapinfo.cpp

```diff
@@ -3356,6 +3360,10 @@ unsigned ZapInfo::getClassNumInstanceFields(CORINFO_CLASS_HANDLE cls)
 CorInfoHelpFunc ZapInfo::getNewHelper(CORINFO_RESOLVED_TOKEN * pResolvedToken, CORINFO_METHOD_HANDLE callerHandle)
 {
     classMustBeLoadedBeforeCodeIsRun(pResolvedToken->hClass);
+
+    if (IsReadyToRunCompilation())
```

Comment:
The early out for R2R should be the very first thing - before classMustBeLoadedBeforeCodeIsRun.